implement enquiry/answerback

jcollie opened the pull request:
Needs a little more testing before merging. One question that I had would it be better for the configuration entry to be called "answerback-string" instead of "enquiry-string".

jcollie committed:
implement enquiry

mitchellh reviewed (commented):
Looks like a good start. The number of allocations required (I think... 3 dupes?) seems highly suspicious to me, but I haven't looked into more detail around lifetimes to think about it carefully. 

I don't see a mechanism for this to reload at runtime safely, as far as I can tell, either.

The core idea seems good.

jcollie committed:
change name of config entry and variables, add lock for safety during config update

jcollie committed:
add some more documentation about enquiry

jcollie committed:
add some more documentation about enquiry on iterm2

mitchellh committed:
termio/exec: remove a bunch of state that is duplicated on StreamHandler

A lot of the state that we put on Exec is just there to copy to
StreamHandler, but we already have it in DerivedConfig. I think this
whole copy copy copy is just legacy cruft since termio.Exec is one of
the older parts of the source code.

This rearchitects the Exec struct to act more like Surface and Renderer
where it stores its derived config. This lets us avoid a few extra
allocations and removes a LOT of struct member noise from termio.Exec.

For pointer lifetimes, the memory allocated is now owned by
DerivedConfig. When changeConfig is called, its the only time BOTH are
still alive, so we can safely swap pointers and deinit without having to
duplicate across threads. This is the same as renderer/surface.

mitchellh committed:
termio/exec: assign arena last so we capture allocations
